feat: add required field support

attiyaIshaque opened the pull request:
Fields can now be added dynamically to the registration page. These changes are currently behind the `ENABLE_DYNAMIC_REGISTRATION_FIELDS `. 



Ticket: https://openedx.atlassian.net/browse/VAN-838

zainab-amir commented on src/register/RegistrationPage.jsx:
```diff
@@ -543,6 +556,44 @@ class RegistrationPage extends React.Component {
 
+    const formFields = Object.keys(this.props.fieldDescriptions).map((fieldName) => {
+      const fieldData = this.props.fieldDescriptions[fieldName];
+      if (fieldData.name === 'country' && fieldData.type === 'select') {
```
As mentioned by Waheed we can remove the second check
```suggestion
      if (fieldData.name === 'country') {
```

zainab-amir commented on src/register/RegistrationPage.jsx:
```diff
@@ -543,6 +556,44 @@ class RegistrationPage extends React.Component {
+        );
+      }
+      if (fieldData.name === 'Honor Code') {
+        return (<HonorCode fieldData={fieldData} value={this.state.values[fieldData.name]} onChangeHandler={this.onChangeHandler} />);
```
nit: If the field is on the same line we don't need to add extra brackets
```suggestion
        return <HonorCode fieldData={fieldData} value={this.state.values[fieldData.name]} onChangeHandler={this.onChangeHandler} />;
```

zainab-amir commented on src/register/data/actions.js:
```diff
@@ -53,3 +54,21 @@ export const fetchRealtimeValidationsFailure = () => ({
+
+export const getRequiredFieldDataSuccess = (data, extendedProfile) => ({
+  type: GET_REQUIRED_FIELDS_DATA.SUCCESS,
+  payload: { data, extendedProfile }, // remove if not required
```
is it required?

zainab-amir commented on src/register/HonorCode.jsx:
```diff
@@ -0,0 +1,71 @@
+        >
+          <FormattedMessage
+            id="register.page.honor.code"
+            defaultMessage="I agree to the {platformName}  {tosAndHonorCode}"
```
nit: extra space
```suggestion
            defaultMessage="I agree to the {platformName} {tosAndHonorCode}"
```

zainab-amir reviewed (changes requested):
I noticed that we are sending error messages from backend but are not using it here for required fields. We should make sure that appropriate error message is shown if required field is not submitted. 

Also please update `handleSubmit` because currently in code these dynamic fields are not being submitted to backend on submission.

shafqatfarhan commented on src/register/RegistrationPage.jsx:
```diff
@@ -627,27 +754,29 @@ class RegistrationPage extends React.Component {
-              />
-            </div>
+            {!(this.showDynamicRegistrationFields) ? (
+              <div id="honor-code" className="micro text-muted mt-4">
+                <FormattedMessage
+                  id="register.page.terms.of.service.and.honor.code"
+                  defaultMessage="By creating an account, you agree to the {tosAndHonorCode} and you acknowledge that {platformName} and each
+                  Member process your personal data in accordance with the {privacyPolicy}."
+                  description="Text that appears on registration form stating honor code and privacy policy"
+                  values={{
+                    platformName: getConfig().SITE_NAME,
+                    tosAndHonorCode: (
+                      <Hyperlink variant="muted" destination={getConfig().TOS_AND_HONOR_CODE || '#'} target="_blank">
+                        {intl.formatMessage(messages['terms.of.service.and.honor.code'])}
+                      </Hyperlink>
+                    ),
+                    privacyPolicy: (
+                      <Hyperlink variant="muted" destination={getConfig().PRIVACY_POLICY || '#'} target="_blank">
+                        {intl.formatMessage(messages['privacy.policy'])}
+                      </Hyperlink>
+                    ),
+                  }}
+                />
```
Can we move this block inside the `HonorCode.jsx` file?

attiyaIshaque replied:
@shafqatfarhan I have changed the code according to your suggestion. please review it.

shafqatfarhan commented on src/field-renderer/FieldRenderer.jsx:
```diff
@@ -6,71 +6,109 @@ import { ExpandMore } from '@edx/paragon/icons';
       formField = (
-        <Form.Group controlId={fieldData.name} className="mb-3">
+        <Form.Group controlId={fieldData.name}>
           <Form.Control
             as="select"
             name={fieldData.name}
             value={value}
             onChange={(e) => onChangeHandler(e)}
             trailingElement={<Icon src={ExpandMore} />}
             floatingLabel={fieldData.label}
+            onBlur={handleOnBlur}
+            onFocus={handleFocus}
           >
```
@attiyaIshaque Please add the `aria-invalid` attribute in all the fields that show error state

Ref: https://github.com/openedx/frontend-app-authn/pull/571/files

attiyaIshaque replied:
Added. :) Thanks for pointing out this.

shafqatfarhan reviewed (commented):
I have noticed a bunch of UI/UX issues:

<img width="540" alt="Screenshot 2022-05-19 at 1 26 33 PM" src="https://user-images.githubusercontent.com/14912524/169271067-4a0271bd-6806-4ca8-9685-204eb2bfa090.png">

- Extra required fields do not have an error (red) border around the field on the active error state

<img width="629" alt="Screenshot 2022-05-19 at 1 38 55 PM" src="https://user-images.githubusercontent.com/14912524/169271714-fb0c81da-f2b1-44fd-b8c8-d595d2c5147e.png">

- When the `ENABLE_DYNAMIC_REGISTRATION_FIELDS` flag is false, the honor code paragraph had spacing (margin-top) with the marketing checkbox which is removed now

attiyaIshaque commented:
@shafqatfarhan I have addressed your comments now
<img width="573" alt="Screenshot 2022-05-19 at 4 02 14 PM" src="https://user-images.githubusercontent.com/7627421/169279414-0ec2ea99-097b-4e58-8039-9831c01caf37.png">

shafqatfarhan reviewed (approved):
LGTM 👍
